p2p/transport/tcp/tcp.go

```diff
@@ -134,15 +148,16 @@ var _ transport.Transport = &TcpTransport{}
 var _ transport.DialUpdater = &TcpTransport{}
 
 // NewTCPTransport creates a tcp transport object that tracks dialers and listeners
-// created. It represents an entire TCP stack (though it might not necessarily be).
-func NewTCPTransport(upgrader transport.Upgrader, rcmgr network.ResourceManager, opts ...Option) (*TcpTransport, error) {
+// created.
+func NewTCPTransport(upgrader transport.Upgrader, rcmgr network.ResourceManager, sharedTCP *tcpreuse.ConnMgr, opts ...Option) (*TcpTransport, error) {
 	if rcmgr == nil {
```

Comment:
This is a breaking change. 
It simplifies the top level basic host api a lot. You can do:

```
libp2p.New(Transport(tcp.NewTransport), Transport(websocket.New), ShareTCPListener())
```
And not need to specify the rcmgr, or the connection gater. 

I'm confused how to provide this API, enabling sharing by just specifying `ShareTCPListener`, and pass in the sharedTCP objects as one of the optional arguments TCP Transport. If we can pass in the sharedTCP object as an optional argument that will prevent breaking this.